compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESimplifyDmaWaits.cpp

```diff
@@ -0,0 +1,158 @@
+LogicalResult simplifyDmaWaits(AMDAIE::AMDAIEDeviceModel deviceModel,
+                               AMDAIE::WorkgroupOp workgroupOp) {
+  IRRewriter rewriter(workgroupOp->getContext());
+  std::vector<AMDAIE::NpuDmaWaitOp> waitOpsToErase;
+  DenseMap<AMDAIE::ConnectionOp, SmallVector<uint32_t>> connectionToBdIdQueues;
+  AMDAIE::ControlCodeOp controlCodeOp = workgroupOp.getControlCode();
+  WalkResult res = controlCodeOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+      [&](AMDAIE::NpuDmaWaitOp waitOp) {
+        bool toErase = true;
+        for (Value token : waitOp.getAsyncTokens()) {
+          if (auto npuHalfDmaCpyNdOp =
+                  dyn_cast_if_present<AMDAIE::NpuHalfDmaCpyNdOp>(
+                      token.getDefiningOp())) {
+            // Retrieve the connection op.
+            std::optional<AMDAIE::ConnectionOp> maybeConnectionOp =
+                npuHalfDmaCpyNdOp.getConnectionOp();
+            if (!maybeConnectionOp) {
+              npuHalfDmaCpyNdOp.emitOpError()
+                  << "expected to operate on an `amdaie.connection`";
+              return WalkResult::interrupt();
+            }
+            AMDAIE::ConnectionOp connectionOp = maybeConnectionOp.value();
+            // Retrieve the flow op.
+            std::optional<AMDAIE::FlowOp> maybeFlowOp =
+                maybeConnectionOp->getFlowOp();
+            if (!maybeFlowOp) {
+              maybeConnectionOp->emitOpError()
+                  << "expected to operate on an `amdaie.flow`";
+              return WalkResult::interrupt();
+            }
+            if (maybeFlowOp->getIsPacketFlow()) return WalkResult::advance();
+            // Retrieve the BD ID op.
+            std::optional<AMDAIE::BdIdOp> maybeBdIdOp =
+                npuHalfDmaCpyNdOp.getBdIdOp();
+            if (!maybeBdIdOp) {
+              npuHalfDmaCpyNdOp.emitOpError()
+                  << "must have a BD ID op to lower to "
+                     "`amdaie.npu.write_bd`";
+              return WalkResult::interrupt();
+            }
+            AMDAIE::BdIdOp bdIdOp = maybeBdIdOp.value();
+            // Retrieve the tile op.
+            AMDAIE::TileOp tileOp = dyn_cast_if_present<AMDAIE::TileOp>(
+                bdIdOp.getTile().getDefiningOp());
+            if (!tileOp) {
+              bdIdOp.emitOpError() << "must operate on an `amdaie.tile`";
+              return WalkResult::interrupt();
+            }
+            // Get the maximum queue size.
+            uint32_t col = getConstantIndexOrAssert(tileOp.getCol());
+            uint32_t row = getConstantIndexOrAssert(tileOp.getRow());
+            uint32_t maxQueueSize = deviceModel.getDmaMaxQueueSize(col, row);
+            // Keep wait op if reaches the maximum queue size or there is a
+            // duplicate BD ID.
+            uint32_t bdId = getConstantIndexOrAssert(bdIdOp.getValue());
+            auto &bdIdQueue = connectionToBdIdQueues[connectionOp];
```

Comment:
```suggestion
            SmallVector<uint32_t> &bdIdQueue = connectionToBdIdQueues[connectionOp];
```
Following LLVM conventions, we avoid `auto` unless the type is clear from the context (`cast<Type>` etc).